homeassistant/components/tuya/__init__.py

```diff
@@ -0,0 +1,296 @@
+            for ha_device in hass.data[DOMAIN][TUYA_HA_DEVICES]:
+                if ha_device.tuya_device.id == device.id:
+                    _LOGGER.debug(
+                        "_update-->%s;->>%s", self, ha_device.tuya_device.status
+                    )
+                    ha_device.schedule_update_ha_state()
+
+        def add_device(self, device: TuyaDevice):
+
+            device_add = False
+
+            _LOGGER.debug(
+                """add device category->%s; keys->,
+                {hass.data[DOMAIN][TUYA_HA_TUYA_MAP].keys()}""",
+                device.category,
+            )
+            if device.category in itertools.chain(
+                *hass.data[DOMAIN][TUYA_HA_TUYA_MAP].values()
+            ):
+                ha_tuya_map = hass.data[DOMAIN][TUYA_HA_TUYA_MAP]
+
+                remove_hass_device(hass, device.id)
+
+                for key, tuya_list in ha_tuya_map.items():
+                    if device.category in tuya_list:
+                        device_add = True
+                        dispatcher_send(
+                            hass, TUYA_DISCOVERY_NEW.format(key), [device.id]
+                        )
+
+            if device_add:
+                device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+                device_manager.mq.stop()
+                tuya_mq = TuyaOpenMQ(device_manager.api)
+                tuya_mq.start()
+
+                device_manager.mq = tuya_mq
+                tuya_mq.add_message_listener(device_manager.on_message)
+
+        def remove_device(self, device_id: str):
+            _LOGGER.debug("tuya remove device:%s", device_id)
+            remove_hass_device(hass, device_id)
+
+    __listener = DeviceListener()
+    hass.data[DOMAIN][TUYA_MQTT_LISTENER] = __listener
+    device_manager.add_device_listener(__listener)
+    hass.data[DOMAIN][TUYA_DEVICE_MANAGER] = device_manager
+
+    # Clean up device entities
+    await cleanup_device_registry(hass)
+
+    _LOGGER.debug("init support type->%s", TUYA_SUPPORT_HA_TYPE)
+
+    for platform in TUYA_SUPPORT_HA_TYPE:
+        _LOGGER.debug("tuya async platform-->%s", platform)
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+        hass.data[DOMAIN][TUYA_SETUP_PLATFORM].add(platform)
+
+    return True
+
+
+async def cleanup_device_registry(hass: HomeAssistant):
+    """Remove deleted device registry entry if there are no remaining entities."""
+
+    __device_registry = device_registry.async_get(hass)
+    device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+
+    for dev_id, device_entity in list(__device_registry.devices.items()):
+        for item in device_entity.identifiers:
+            if DOMAIN == item[0] and item[1] not in device_manager.device_map:
+                __device_registry.async_remove_device(dev_id)
+                break
+
+
+def remove_hass_device(hass: HomeAssistant, device_id: str):
```

Comment:
Mark this method as a callback:

```suggestion
from homeassistant.core import callback

@callback
def async_remove_hass_device(hass: HomeAssistant, device_id: str):
```

Then when you want to call it from a sync context:

```python
hass.add_job(async_remove_hass_device, hass, device_id)
```

Reply:
fixed